File: neuro-backend/src/domain/entities/kanban.rs
//! =============================================================================
//! Kanban Domain Entities
//! =============================================================================

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// =============================================================================
// Card (Tarjeta Kanban)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Card {
    pub id: Uuid,
    pub column_id: Uuid,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    pub labels: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due_date: Option<DateTime<Utc>>,
    #[serde(rename = "card_order")]
    pub order: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateCard {
    pub title: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub labels: Option<Vec<String>>,
    #[serde(default)]
    pub due_date: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateCard {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub labels: Option<Vec<String>>,
    #[serde(default)]
    pub due_date: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MoveCard {
    pub target_column_id: Uuid,
    pub target_order: i32,
}

// =============================================================================
// Column (Columna Kanban)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Column {
    pub id: Uuid,
    pub board_id: Uuid,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wip_limit: Option<i32>,
    #[serde(rename = "column_order")]
    pub order: i32,
    pub cards: Vec<Card>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateColumn {
    pub name: String,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub wip_limit: Option<i32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateColumn {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub wip_limit: Option<i32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ReorderColumn {
    pub target_order: i32,
}

// =============================================================================
// Board (Tablero Kanban)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Board {
    pub id: Uuid,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    pub is_archived: bool,
    pub columns: Vec<Column>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BoardSummary {
    pub id: Uuid,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub color: Option<String>,
    pub is_archived: bool,
    pub column_count: usize,
    pub card_count: usize,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateBoard {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    /// If true, create default columns (To Do, In Progress, Done)
    #[serde(default)]
    pub with_default_columns: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateBoard {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub color: Option<String>,
    #[serde(default)]
    pub is_archived: Option<bool>,
}

impl Board {
    pub fn to_summary(&self) -> BoardSummary {
        let card_count: usize = self.columns.iter().map(|c| c.cards.len()).sum();
        BoardSummary {
            id: self.id,
            name: self.name.clone(),
            description: self.description.clone(),
            color: self.color.clone(),
            is_archived: self.is_archived,
            column_count: self.columns.len(),
            card_count,
            created_at: self.created_at,
            updated_at: self.updated_at,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_board() -> Board {
        Board {
            id: Uuid::new_v4(),
            name: "Test Board".to_string(),
            description: None,
            color: None,
            is_archived: false,
            columns: vec![],
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }

    #[test]
    fn test_board_to_summary_empty() {
        let board = make_board();
        let summary = board.to_summary();
        assert_eq!(summary.column_count, 0);
        assert_eq!(summary.card_count, 0);
        assert_eq!(summary.name, "Test Board");
    }

    #[test]
    fn test_board_to_summary_with_cards() {
        let mut board = make_board();
        let col_id = Uuid::new_v4();
        let card = Card {
            id: Uuid::new_v4(),
            column_id: col_id,
            title: "Task".to_string(),
            description: None,
            color: None,
            labels: vec![],
            due_date: None,
            order: 0,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        };
        board.columns.push(Column {
            id: col_id,
            board_id: board.id,
            name: "To Do".to_string(),
            color: None,
            wip_limit: None,
            order: 0,
            cards: vec![card],
            created_at: Utc::now(),
            updated_at: Utc::now(),
        });
        let summary = board.to_summary();
        assert_eq!(summary.column_count, 1);
        assert_eq!(summary.card_count, 1);
    }

    #[test]
    fn test_move_card_fields() {
        let mv = MoveCard {
            target_column_id: Uuid::new_v4(),
            target_order: 2,
        };
        assert_eq!(mv.target_order, 2);
    }

    #[test]
    fn test_create_board_with_defaults() {
        let req = CreateBoard {
            name: "Sprint".to_string(),
            description: None,
            color: Some("#ff0000".to_string()),
            with_default_columns: Some(true),
        };
        assert_eq!(req.name, "Sprint");
        assert_eq!(req.with_default_columns, Some(true));
    }

    #[test]
    fn test_update_board_partial() {
        let req = UpdateBoard {
            name: None,
            description: Some("Updated".to_string()),
            color: None,
            is_archived: Some(true),
        };
        assert_eq!(req.is_archived, Some(true));
        assert!(req.name.is_none());
    }
}

File: neuro-ui/src/api/kanban.ts
import api from './client';

export interface KanbanCardDto {
  id: string;
  column_id: string;
  title: string;
  description?: string;
  color?: string;
  labels: string[];
  due_date?: string;
  order: number;
  created_at: string;
  updated_at: string;
}

export interface KanbanColumnDto {
  id: string;
  board_id: string;
  name: string;
  color?: string;
  wip_limit?: number;
  order: number;
  cards: KanbanCardDto[];
  created_at: string;
  updated_at: string;
}

export interface KanbanBoardDto {
  id: string;
  name: string;
  description?: string;
  color?: string;
  is_archived: boolean;
  columns: KanbanColumnDto[];
  created_at: string;
  updated_at: string;
}

export interface KanbanBoardSummaryDto {
  id: string;
  name: string;
  description?: string;
  color?: string;
  is_archived: boolean;
  column_count: number;
  card_count: number;
  created_at: string;
  updated_at: string;
}

export interface CreateBoardRequest {
  name: string;
  description?: string;
  color?: string;
  with_default_columns?: boolean;
}

export interface UpdateBoardRequest {
  name?: string;
  description?: string;
  color?: string;
  is_archived?: boolean;
}

export interface CreateColumnRequest {
  name: string;
  color?: string;
  wip_limit?: number;
}

export interface UpdateColumnRequest {
  name?: string;
  color?: string;
  wip_limit?: number;
}

export interface CreateCardRequest {
  title: string;
  description?: string;
  color?: string;
  labels?: string[];
  due_date?: string;
}

export interface UpdateCardRequest {
  title?: string;
  description?: string;
  color?: string;
  labels?: string[];
  due_date?: string;
}

export interface MoveCardRequest {
  target_column_id: string;
  target_order: number;
}

export const kanbanApi = {
  listBoards: async (): Promise<KanbanBoardSummaryDto[]> => {
    const response = await api.get<KanbanBoardSummaryDto[]>('/kanban/boards');
    return response.data;
  },

  getBoard: async (boardId: string): Promise<KanbanBoardDto> => {
    const response = await api.get<KanbanBoardDto>(`/kanban/boards/${boardId}`);
    return response.data;
  },

  createBoard: async (request: CreateBoardRequest): Promise<KanbanBoardDto> => {
    const response = await api.post<KanbanBoardDto>('/kanban/boards', request);
    return response.data;
  },

  updateBoard: async (boardId: string, request: UpdateBoardRequest): Promise<KanbanBoardDto> => {
    const response = await api.patch<KanbanBoardDto>(`/kanban/boards/${boardId}`, request);
    return response.data;
  },

  deleteBoard: async (boardId: string): Promise<void> => {
    await api.delete(`/kanban/boards/${boardId}`);
  },

  createColumn: async (boardId: string, request: CreateColumnRequest): Promise<KanbanColumnDto> => {
    const response = await api.post<KanbanColumnDto>(`/kanban/boards/${boardId}/columns`, request);
    return response.data;
  },

  updateColumn: async (boardId: string, columnId: string, request: UpdateColumnRequest): Promise<KanbanColumnDto> => {
    const response = await api.patch<KanbanColumnDto>(`/kanban/boards/${boardId}/columns/${columnId}`, request);
    return response.data;
  },

  deleteColumn: async (boardId: string, columnId: string): Promise<void> => {
    await api.delete(`/kanban/boards/${boardId}/columns/${columnId}`);
  },

  createCard: async (boardId: string, columnId: string, request: CreateCardRequest): Promise<KanbanCardDto> => {
    const response = await api.post<KanbanCardDto>(`/kanban/boards/${boardId}/columns/${columnId}/cards`, request);
    return response.data;
  },

  updateCard: async (boardId: string, columnId: string, cardId: string, request: UpdateCardRequest): Promise<KanbanCardDto> => {
    const response = await api.patch<KanbanCardDto>(`/kanban/boards/${boardId}/columns/${columnId}/cards/${cardId}`, request);
    return response.data;
  },

  deleteCard: async (boardId: string, columnId: string, cardId: string): Promise<void> => {
    await api.delete(`/kanban/boards/${boardId}/columns/${columnId}/cards/${cardId}`);
  },

  moveCard: async (boardId: string, columnId: string, cardId: string, request: MoveCardRequest): Promise<KanbanBoardDto> => {
    const response = await api.put<KanbanBoardDto>(`/kanban/boards/${boardId}/columns/${columnId}/cards/${cardId}/move`, request);
    return response.data;
  },
};
